src/main/java/org/broadinstitute/hellbender/tools/spark/sv/InsertSizeMetricsCollectorSpark.java

```diff
@@ -147,48 +300,182 @@ private void collectSymmetricalBinWidth(final JavaRDD<Integer> lengthRDD){
-     * Mean and SD are severely impacted by outliers, so they are computed here after trimming.
-     * @param HistogramMADTolerance how many MADs from the median will be tolerated when down-trimming
-     */
-    private void trimHistogramAndCollectMeanAndSD(final JavaRDD<Integer> lengthRDD, final double HistogramMADTolerance){
+    // small utility function to trim htsjdk Histogram and set corresponding metric's mean and SD
+    private static void trimHTSHistogramAndSetMean(Histogram<Integer> htsHist, InsertSizeMetrics metrics, final double histogramMADTolerance){
+        htsHist.trimByWidth( (int)(metrics.MEDIAN_INSERT_SIZE + histogramMADTolerance*metrics.MEDIAN_ABSOLUTE_DEVIATION) );
+        metrics.MEAN_INSERT_SIZE   = htsHist.getMean();
+        if(1==htsHist.getCount()){ // extremely unlikely in reality, but may be true at read group level when running tests
+            metrics.STANDARD_DEVIATION = 0.0;
+        }else{
+            metrics.STANDARD_DEVIATION = htsHist.getStandardDeviation();
+        }
+    }
 
-        // range of read length that is going to be tolerated when computing histogram, mean and SD, but will not affect other metrics.
-        final int lengthToleranceLowerEnd = (int) metrics.MEDIAN_INSERT_SIZE - (int) (HistogramMADTolerance * metrics.MEDIAN_ABSOLUTE_DEVIATION);
-        final int lengthToleranceHigherEnd = (int) metrics.MEDIAN_INSERT_SIZE + (int) (HistogramMADTolerance * metrics.MEDIAN_ABSOLUTE_DEVIATION);
-        final JavaRDD<Integer> trimmedLengthRDD = lengthRDD.filter(length -> length >= lengthToleranceLowerEnd)
-                                                           .filter(length -> length <= lengthToleranceHigherEnd);
-        final double lengthArr[] = getRDDArray(trimmedLengthRDD);
+    // small utility function to decide what group name to use in the corresponding htsjdk Histogram title/ctor.
+    private static String getGroupName(final GroupInfo groupInfo){
+        String groupName = null;
+        switch (groupInfo.level){
+            case ALL_READS:
+                groupName = ALL_READS_KEY_MACRO;
+                break;
+            case SAMPLE:
+                groupName = groupInfo.sample;
+                break;
+            case LIBRARY:
+                groupName = groupInfo.library;
+                break;
+            case READ_GROUP:
+                groupName = groupInfo.readGroup;
+                break;
+        }
+        return (groupName + ".");
+    }
+
+    private static String orientationToString(final SamPairUtil.PairOrientation OR){
+        return OR.equals(SamPairUtil.PairOrientation.FR) ? "fr" : (OR.equals(SamPairUtil.PairOrientation.RF) ? "rf" : "tandem");
+    }
+
+    // utility functions to do mapping on RDDs; broken into three steps for easier comprehension
+    private static Tuple2<GroupInfo, Tuple2<SamPairUtil.PairOrientation, Integer>> step1ReadTraversal(final GATKRead read, final SAMFileHeader header){
```

Comment:
Again give this and the following functions a better name, ie "mapReadToGroupInfoOrientationAndInsertSize"